aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,127 @@
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Populate `num_batches` number of torch::jit::Stack, one for each computation.
+  std::vector<torch::jit::Stack> unbatched_stacks(num_batches);
+  auto pushToEachStack = [&](const auto& ivalue) {
+    for (auto& stack : unbatched_stacks) {
+      torch::jit::push(stack, ivalue);
+    }
+  };
+  auto batched_tensor_inputs_pos_iter = batched_tensor_inputs_position.begin();
+	auto input_physical_views_iter = input_physical_views.begin();
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
+    if (idx != *batched_tensor_inputs_pos_iter) {
+      // ivalue isn't a BatchedTensor
+      pushToEachStack(ivalue);
+      continue;
+    }
+    // ivalue is a BatchedTensor
+    const auto& physical_view_for_ivalue = *input_physical_views_iter;
+    for (int64_t linear_idx = 0; linear_idx < num_batches; ++linear_idx) {
+      auto index = computeIndex(linear_idx, batch_sizes);
+      torch::jit::push(
+          unbatched_stacks[linear_idx],
+          physical_view_for_ivalue.tensor().index(index));
+    }
+    batched_tensor_inputs_pos_iter++;
+    input_physical_views_iter++;
+  }
+
+  // Call the operation once for batch
+  for (auto& stack : unbatched_stacks) {
+    op.callBoxed(&stack);
+  }
+
+  // Stack the tensors together to form the result.
+  stack->clear();
+  std::vector<Tensor> output_shards;
+  for (const auto& stack : unbatched_stacks) {
+    TORCH_INTERNAL_ASSERT(stack.size() == 1)
+    output_shards.push_back(stack[0].toTensor());
+  }
+  auto flat_output = at::stack(output_shards);
```

Comment:
The `stack` means that the fallback kernel won't be as performant as a user deciding to use a for-loop rather than vmap. However, the long-term state for vmap is to not rely on the fallback (and instead have batching rules defined on the operations) so I am not too worried about the performance.

## Problem

Consider the following model:
```
def model(example):
    example = F.conv(example, ...)
    example = F.relu(example, ...)
    example = F.linear(example, ...)
    return tensor
```
If the user wanted to push multiple examples via a Python for-loop, they’d do the following:
```
examples = torch.randn(N, ...)
results = []
for example in examples.unbind(0):
    results.append(model(example))
output = torch.stack(results)
```
Now, let’s assume that no batching rules were implemented for F.conv, F.relu, and F.linear. The vmap “for-loop” fallback would do the following, for each operation:

* unbind the examples along the batch dimension
* run the operator for each example
* stack the results

In other words, semantically, vmap(model)(examples) is equivalent to:
```
examples = torch.stack([F.conv(example, ...) for example in examples])
examples = torch.stack([F.relu(example, ...) for example in examples])
examples = torch.stack([F.linear(example, ...) for example in examples])
```
All of the torch.stack make the vmap “for-loop” fallback slower than the python for-loop, which is not very good for vmap.

Reply:
The obvious fix is to not stack: instead, use `out` to write samples directly into a preallocated output. But to preallocate the output, you need a way of computing what the size of the computation will be on one output, which is not so easy to do right now. One hacky way to do it would be to run on one example, then resize the output so it's large enough to fit the rest of the examples, and then run the loop. It'll be faster than stack. Then if we have working meta functions, you can use those to compute what the output size should be after the fact.

Reply:
We can't actually use the `out=` variant of the function because they doesn't support autograd. Because vmap occurs above autograd in the current design, vmap must desugar into operations that are compatible with autograd.

Reply:
RIP. cc @albanD

Reply:
Note that this will depend on the overall model: a short model (few ops) won't see much difference while a deeper model will be more impacted.
Also I guess that the output size vs complexity of the function itself will have a big impact here.

But in general, for functions with inplace versions (`foo_()`), we could do this and keep autograd. But we will need a list of such functions and register a different fallback kernel for these.
For other functions, I can't think of a better way to do it :/

Reply:
@albanD what is being asked for here is a little different. Imagine:

```
x1 = torch.randn(2, requires_grad=True)
x2 = torch.randn(2, requires_grad=True)
x3 = torch.randn(2, requires_grad=True)
y = torch.empty(3, 2)
x1.add(w, out=y[0])
x2.add(w, out=y[1])
x3.add(w, out=y[2])
y.backward()
```

It would be cool if this worked.

Reply:
Yes it would :D 
As mentioned in the `out=` autograd proposal [here](https://github.com/pytorch/pytorch/issues/29378#issuecomment-571805569), that would work just fine.
Unfortunately, detecting these valid cases and implementing this overall is not going to be done any time soon I think.

The version for the `foo_()` variants would work with the current code base though (even though it won't work with as many ops).